paasd/src/store.rs

```diff
@@ -0,0 +1,97 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+
+use uuid::Uuid;
+
+use worker::Process;
+
+use crate::user::UserId;
+
+#[derive(Debug, thiserror::Error)]
+pub enum GetError {
+    #[error("Process with the given id not found")]
+    NotFound,
+    #[error("User unauthorized to access this process")]
+    Unauthorized,
+}
+
+impl Into<tonic::Status> for GetError {
+    fn into(self) -> tonic::Status {
+        use GetError::*;
+        match self {
+            NotFound => tonic::Status::not_found(format!("{}", self)),
+            // TODO: unauthenticated != unauthorized, better Status?
+            Unauthorized => tonic::Status::unauthenticated(format!("{}", self)),
+        }
+    }
+}
+
+pub type ProcessStore = Store<Process>;
+
+type StoreMap<V> = HashMap<Uuid, (UserId, Arc<V>)>;
+
+#[derive(Default)]
+pub struct Store<V>(RwLock<StoreMap<V>>);
+
+impl<V> Store<V> {
+    pub fn new() -> Self {
+        Self(RwLock::new(StoreMap::new()))
+    }
+
+    pub fn get(&self, pid: Uuid, uid: &UserId) -> Result<Arc<V>, GetError> {
+        let read = self.0.read().unwrap();
+        let (owner, process) = read.get(&pid).ok_or(GetError::NotFound)?;
+        // TODO: does this need constant-time EQ?
+        if owner == uid {
+            Ok(process.clone())
+        } else {
+            Err(GetError::Unauthorized)
```

Comment:
This allows any user to find out information about jobs from other users since you differentiate between `NotFound`and `Unauthorized`. Generally giving out any more information that is needed about other users is a liability and security risk, akin to how you can often check if accounts are registered on websites with certain emails.

Reply:
Good point. I actually thought about this, but forgot to leave a comment about the UX vs security tradeoff. Will fix.